### PR TITLE
Use leader election mechanism from controller-runtime library

### LINKED ISSUE #295: Leader election does not work correct after node power off
## Environment

```
# oc get nodes
NAME       STATUS   ROLES           AGE   VERSION
master-0   Ready    master,worker   21h   v1.14.0+5ccbd432f
master-1   Ready    master,worker   21h   v1.14.0+5ccbd432f
master-2   Ready    master,worker   21h   v1.14.0+5ccbd432f

# oc get pods -n openshift-machine-api -o wide
NAME                                           READY   STATUS    RESTARTS   AGE   IP               NODE       NOMINATED NODE   READINESS GATES
cluster-autoscaler-operator-6b5957fbd8-zcwm4   1/1     Running   2          20h   10.129.0.59      master-2   <none>           <none>
machine-api-controllers-6b79798968-bqmj4       3/3     Running   0          17h   10.128.0.56      master-1   <none>           <none>
machine-api-operator-857dd85bb9-8jcfm          1/1     Running   0          17h   10.128.0.57      master-1   <none>           <none>
metal3-7685f8bdc9-r7zz2                        8/8     Running   0          17h   192.168.111.22   master-2   <none>           <none>

# oc get bmh --namespace openshift-machine-api
NAME                 STATUS   PROVISIONING STATUS      CONSUMER        BMC                         HARDWARE PROFILE   ONLINE   ERROR
openshift-master-0   OK       externally provisioned   demo-master-0   ipmi://192.168.111.1:6230                      true     
openshift-master-1   OK       externally provisioned   demo-master-1   ipmi://192.168.111.1:6231                      true     
openshift-master-2   OK       externally provisioned   demo-master-2   ipmi://192.168.111.1:6232                      true
```

## How to reproduce

1. Check where the BMO pod is running, in our case, it is running on the `master-2`
2. Power-off the `master-2` node via setting `spec.Online: false`, `# oc -n openshift-machine-api edit bmh openshift-master-2`
3. Verify that node does not run anymore under the libvirt
```
# virsh list
 Id    Name                           State
----------------------------------------------------
 10    openshift_master_1             running
 13    openshift_master_0             running
```
4. Verify that the pod restarted
```
# oc get pods -n openshift-machine-api -o wide
NAME                                           READY   STATUS        RESTARTS   AGE   IP               NODE       NOMINATED NODE   READINESS GATES
cluster-autoscaler-operator-6b5957fbd8-zcwm4   1/1     Running       2          21h   10.129.0.59      master-2   <none>           <none>
machine-api-controllers-6b79798968-bqmj4       3/3     Running       0          17h   10.128.0.56      master-1   <none>           <none>
machine-api-operator-857dd85bb9-8jcfm          1/1     Running       0          17h   10.128.0.57      master-1   <none>           <none>
metal3-7685f8bdc9-5nd65                        8/8     Running       0          69s   192.168.111.20   master-0   <none>           <none>
metal3-7685f8bdc9-r7zz2                        8/8     Terminating   0          17h   192.168.111.22   master-2   <none>           <none>
```
5. Verify that new pod fails to get leader election
```
# oc -n openshift-machine-api logs metal3-7685f8bdc9-5nd65 -c metal3-baremetal-operator
{"level":"info","ts":1567497633.7019377,"logger":"cmd","msg":"Go Version: go1.11.6"}
{"level":"info","ts":1567497633.7019632,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1567497633.7019677,"logger":"cmd","msg":"Version of operator-sdk: v0.8.0+git"}
{"level":"info","ts":1567497633.7021453,"logger":"cmd","msg":"Component version: metal3-io/baremetal-operator was not built with version info"}
{"level":"info","ts":1567497633.7024767,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1567497699.3227773,"logger":"leader","msg":"Found existing lock","LockOwner":"metal3-7685f8bdc9-r7zz2"}
{"level":"info","ts":1567497699.3331923,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497700.4640396,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497702.8499641,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497707.3879468,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497716.0989492,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497733.4708111,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497751.678254,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497767.9052875,"logger":"leader","msg":"Not the leader. Waiting."}
{"level":"info","ts":1567497784.4164011,"logger":"leader","msg":"Not the leader. Waiting."}
...
```

### PR DESCRIPTION
**Pull request by cynepco3hahue**
There are[ two options for the leader election](https://docs.openshift.com/container-platform/4.1/applications/operator_sdk/osdk-leader-election.html). The SDK defaults to the simpler one ("Leader for life") which trades off slower failover (5 minutes by default - the pod-eviction-timeout) for a guarantee that there can never be two leaders. The other one (used here) fails over faster, but there may be two leaders if one node is running a clock 50% faster(!) than the others unless you tweak the default timeouts to accommodate the clock skew rate you expect.

The controller-runtime library provides a way to set leader among pods via config map annotation, that includes leader lease. In case when the pod fails to update the lease under the config map, the other pod will acquire the leader.

Otherwise, the operator-SDK do not provide the lease feature and just creates the
config map, so until the config map exists, another pod can not acquire the leader,
but the config map will exist as long as pod exists(also in terminating state),
so it will prevent HA in cases when node not responsive for some reason(network, kernel panic, ...).

Fixes #295

### CODE REVIEW

**Comment by nordixinfra** [comment]
Can one of the admins verify this patch?

**Comment by cynepco3hahue** [comment]
@dhellmann Can you please review it?
And do I need to do something to cherry-pick it under the release-4.2 branch?

**Comment by cynepco3hahue** [comment]
@dhellmann You can find the whole information under the description
Fixes #295

**Comment by hardys** [comment]
> @dhellmann You can find the whole information under the description
> Fixes #295

That issue contains a reproducer and demonstrates the symptoms, but doesn't answer any of the questions from @dhellmann AFAICS?

**Comment by cynepco3hahue** [comment]
@hardys @dhellmann Will try to explain better how the leader election is working, in case of controller-runtime it creates separate configmap for each component that uses leader election and updates the annotation of this config map with relevant data(current leader, last lease update time...), and it works perfectly for other repositories, regarding the `operator-sdk` leader election, it just creates config map without any annotations(and I think it just check the existence of this config map), I do not sure if it some operator-sdk version bug and I do not have time to check it, but it hardly affect HA of the BMO.
Give me know if you need more information.

**Comment by dhellmann** [comment]
> @hardys @dhellmann Will try to explain better how the leader election is working, in case of controller-runtime it creates separate configmap for each component that uses leader election and updates the annotation of this config map with relevant data(current leader, last lease update time...), and it works perfectly for other repositories, regarding the `operator-sdk` leader election, it just creates config map without any annotations(and I think it just check the existence of this config map), I do not sure if it some operator-sdk version bug and I do not have time to check it, but it hardly affect HA of the BMO.
> Give me know if you need more information.

The commit messages for patches should explain why the change is needed. "Fixes a bug" is not enough of a description, for me. Please add some information to the commit message explaining the original problem and how this change addresses it.

I'll also say that if we don't understand why this change fixes anything, then I'm not comfortable merging it. I have not observed any issues with stopping the operator and starting a new one. It's not clear from the original bug report that you waited long enough for the cluster to recover from a node being powered off uncleanly, which I would expect to take longer than a clean shutdown and restart.

**Comment by cynepco3hahue** [comment]
> The commit messages for patches should explain why the change is needed. "Fixes a bug" is not enough of a description, for me. Please add some information to the commit message explaining the original problem and how this change addresses it.

I will update the commit message.

> I'll also say that if we don't understand why this change fixes anything, then I'm not comfortable merging it. I have not observed any issues with stopping the operator and starting a new one. It's not clear from the original bug report that you waited long enough for the cluster to recover from a node being powered off uncleanly, which I would expect to take longer than a clean shutdown and restart.

@dhellmann Did you try to shut down the node where you run the BMO pod? Or you just deleted the BMO pod?
These scenarios are very different from the k8s point of view.

**Comment by zaneb** [comment]
OK, once I knew where to look this is actually quite [well-documented](https://docs.openshift.com/container-platform/4.1/applications/operator_sdk/osdk-leader-election.html).

TL;DR There are two options for leader election. The SDK defaults to the simpler one ("Leader for life") which trades off slower failover (5 minutes by default - the pod-eviction-timeout) for a guarantee that there can never be two leaders. The other one (used here) fails over faster, but there may be two leaders if one node is running a clock 50% faster(!) than the others, unless you tweak the default timeouts to accomodate the clock skew rate you expect.

The only circumstances in which I could imagine this being of even slight concern would be where the operator is running on a VM that gets suspended and then resumed. I'm not even convinced that can cause a problem with the second method that wouldn't also apply to the first. Given that the purpose of this operator is to manage baremetal and therefore it's almost always going to run on baremetal, I don't think even that scenario is worth worrying about.

**Comment by dhellmann** [comment]
> OK, once I knew where to look this is actually quite [well-documented](https://docs.openshift.com/container-platform/4.1/applications/operator_sdk/osdk-leader-election.html).
> 
> TL;DR There are two options for leader election. The SDK defaults to the simpler one ("Leader for life") which trades off slower failover (5 minutes by default - the pod-eviction-timeout) for a guarantee that there can never be two leaders. The other one (used here) fails over faster, but there may be two leaders if one node is running a clock 50% faster(!) than the others, unless you tweak the default timeouts to accomodate the clock skew rate you expect.
> 
> The only circumstances in which I could imagine this being of even slight concern would be where the operator is running on a VM that gets suspended and then resumed. I'm not even convinced that can cause a problem with the second method that wouldn't also apply to the first. Given that the purpose of this operator is to manage baremetal and therefore it's almost always going to run on baremetal, I don't think even that scenario is worth worrying about.

Thank you, @zaneb , that's very clear.

@cynepco3hahue could you put some of these details (including the doc link) in the commit message? I think I'm comfortable merging it once we have a good documentation trail in place.

**Comment by cynepco3hahue** [comment]
@zaneb Thanks for the clarification and for documentation link.
@dhellmann Updated the PR description.

**Comment by dhellmann** [comment]
Thanks @cynepco3hahue
